### PR TITLE
Azure mysql postgres auto discovery configuration

### PR DESCRIPTION
**Pull request by GavinFrazar**
This PR adds configuration for Azure database auto-discovery.

This is the first of several PRs to implement #14688

Configuration allows a user to filter by subscription, resource group, region, tags. If subscription or resource group is omitted, then all subscriptions or all resource_groups are queried. Users can also use `*` to specify this behavior explicitly. Example configs:

```
db_service:
  enabled: "yes"
  resources:
    - labels:
        "*": "*"
  azure:
    # discover all postgres DBs in all subscriptions and resource groups, in the eastus region.
    - subscriptions: ["*"]
      resource_groups: ["*"]
      types: ["postgres"]
      regions: ["eastus"]
      tags:
        "*": "*"
    # discover all mysql DBs in all subscriptions and resource groups, in the eastus or centralus.
    - types: ["mysql"]
      regions: ["eastus", "centralus"]
      tags:
        "*": "*"
    # discover all mysql and postgres DBs a specific subscription and resource group, in the eastus region
    - subscriptions: ["060a97ea-3a57-4218-9be5-dba3f19ff2b5"]
      resource_groups: ["database-access"]
      types: ["mysql", "postgres"]
      regions: ["eastus"]
      tags:
        "*": "*"
```

### CODE REVIEW

**Comment by GavinFrazar** [comment]
Force pushed to split out some unneeded code and restructure my commits.

File: lib/services/matchers.go
Comment (Tener) [inline lib/services/matchers.go:51]:
```suggestion
	// Types are Azure database types to match, for example "mysql" or "postgres".
```

File: lib/config/configuration.go
Comment (Tener) [inline lib/config/configuration.go:1145]: This fixup is a bit strange; I'd expect this logic of equivalence between `len(matcher.Subscriptions) == 0` and `matcher.Subscriptions = []string{types.Wildcard}` to be part of core matching logic, but that is not part of this PR. Other matchers, AWS in particular, do not have this sort of logic applied here, which I think is a good choice.

**Review by Tener — APPROVED** [review]
As far as the configuration goes, apart from the small comments, I think this looks good.

It would be good to see the larger context still, i.e., actual db watcher code that would consume this configuration and execute the matching logic itself.

File: lib/config/configuration_test.go
Comment (smallinsky) [inline lib/config/configuration_test.go:820]: The the `cmp.AllowUnexported(Service{})))` option is needed ? We are compering the AzureMatchers so  `AllowUnexported(Service{})` looks superfluous

File: lib/config/fileconf.go
Comment (smallinsky) [inline lib/config/fileconf.go:1199]: For consistency :-) 
```suggestion
	// Types are Azure database types to match: "mysql", "postgres"
```

File: lib/config/fileconf.go
Comment (smallinsky) [inline lib/config/fileconf.go:1195]:
```suggestion
	// Subscriptions are Azure subscriptions to query for resources.
```

File: lib/services/matchers.go
Comment (smallinsky) [inline lib/services/matchers.go:47]: nit:
```suggestion
	// Subscriptions are Azure subscriptions to query for resources.
```

File: lib/config/configuration.go
Comment (greedy52) [reply lib/config/configuration.go:1145]: The **Contains** check seems an optimization. I don't think we have done it for AWS Matchers (for ex: deduplicate regions). But I also agree that this logic belongs where matching happens.

File: lib/config/configuration_test.go
Comment (greedy52) [inline lib/config/configuration_test.go:425]: In real life, `rg1` would belong to only one of `sub1` or `sub2` (or maybe none for them). I will see how the implementation works out on this.

File: lib/config/fileconf.go
Comment (greedy52) [inline lib/config/fileconf.go:1205]: Both `Subscriptions` and `ResourceGroups` support tags. I assume these `Tags` are only for resources?

File: lib/config/testdata_test.go
Comment (greedy52) [inline lib/config/testdata_test.go:183]: Ignore this comment if you are planning to update `databaseAgentConfigurationTemplate` in future PRs.

File: lib/config/fileconf.go
Comment (greedy52) [reply lib/config/fileconf.go:1205]: I couldn't find an RFD on this. Just curious if we have discussed whether we want to use tags to filter subscriptions and resource groups vs using explicit names.

**Comment by codingllama** [comment]
Just dropping by to say that the PR sizes are excellent now, thanks for the refactors @GavinFrazar.  🤌

You've got a healthy number of active reviewers, so I'll leave it to them.

File: lib/config/configuration.go
Comment (GavinFrazar) [reply lib/config/configuration.go:1145]: The larger context PR was enormous so I split it up.

I think this fixup is about configuration, not about matching logic. The contains check is indeed an optimization, and I should also include deduplication, but I think if two configurations are equivalent it makes sense to simplify the config as soon as possible. It's also much simpler to write and read tests for configuration equivalence than for matching behavior equivalence. I don't feel too strongly either way though, so I'm open to moving it if you still disagree.

File: lib/config/configuration_test.go
Comment (GavinFrazar) [reply lib/config/configuration_test.go:820]: It is superfluous, I'm not sure how it slipped in there. Removing

File: lib/config/configuration_test.go
Comment (GavinFrazar) [reply lib/config/configuration_test.go:425]: It is as you noted the same scenario as when the resource group is not in the subscription at all. There is one api client per `Subscription x DB Type` and a fetcher is made for each tuple `Matcher x Subscription x DB Type x Resource Group`, using the appropriate api client for the subscription and DB type. When the fetcher tries to query for a resource group that does not exist in its subscription, it gets a "Not Found" error and that fetcher returns no results.

File: lib/config/fileconf.go
Comment (GavinFrazar) [reply lib/config/fileconf.go:1205]: Tags are for resources, yes. I'll rename to "ResourceTags". We haven't discussed this anywhere before, and it actually didn't occur to me until you pointed it out. You can specify `*` for subscriptions/groups to get the behavior of only using resource tags for filtering. Or you can be more specific about which subscriptions/groups you want to query. This already enables more specific queries than we support for AWS.

Looking at the SDK, it would require quite a bit more code to watch for subscriptions and resource groups by tag, and then spin up watchers for resources within those subscriptions and groups. Not only that, but there is no API query for finding the subscriptions and resource groups by tag - you would need to query for all of them and then filter the results by tag.

File: lib/config/testdata_test.go
Comment (GavinFrazar) [reply lib/config/testdata_test.go:183]: Not sure what you mean here?

Comment (greedy52) [reply lib/config/testdata_test.go:183]: sorry. here is the link
https://github.com/gravitational/teleport/blob/5d9aa74375ae64cf2205993b09f80edd0bab16fe/lib/config/database.go#L36-L45

File: lib/config/fileconf.go
Comment (greedy52) [reply lib/config/fileconf.go:1205]: Sounds good. We can keep it simple for now.

**Review by greedy52 — APPROVED** [review]

File: lib/config/configuration_test.go
Comment (greedy52) [reply lib/config/configuration_test.go:425]: That makes sense. Thanks for explaining.

File: lib/config/testdata_test.go
Comment (GavinFrazar) [reply lib/config/testdata_test.go:183]: Oh, yes that's in another PR: #15630

**Comment by GavinFrazar** [comment]
@smallinsky friendly ping for re-review, since I need approval from both reviewers list. I think I addressed all of your concerns.

File: lib/config/configuration.go
Comment (Tener) [reply lib/config/configuration.go:1147]: My concern is this code, which defines the semantics of empty resource group match, is not guaranteed to be applied consistently in all cases or be kept in the future.

For example, I expect the `AzureMatcher` to be used in the future for matching VMs and not just dbs. The person implementing that would have to copy these few lines. If they fail to do so, the behavior of matchers will diverge. 

In general, it is hard to ensure how the `AzureMatcher` struct will be constructed in the future; it makes sense to ensure that it behaves in the same way independent of the way that happened.

**Review by smallinsky — APPROVED** [review]

File: lib/config/configuration.go
Comment (GavinFrazar) [reply lib/config/configuration.go]: Ok I'll move it to the code using the matchers. I agree that locality is good - keep code that affects other code closer together. However, this logic will still need to be duplicated for azure VMs or other resources.

**Pull request merged**
